static/js/upload.js

```diff
@@ -13,187 +17,223 @@ exports.feature_check = function (upload_button) {
+            compose_ui.autosize_textarea();
+            uppy.addFile({
+                source: exports.get_item("source", config),
+                name: file.name,
+                type: file.type,
+                data: file,
+            });
+        } catch (error) {
+            // Errors are handled by info-visible and upload-error event callbacks.
         }
+    }
+};
 
-        switch (error_code) {
-        case 'BrowserNotSupported':
-            msg = i18n.t("File upload is not yet available for your browser.");
-            break;
-        case 'TooManyFiles':
-            msg = i18n.t("Unable to upload that many files at once.");
-            break;
-        case 'FileTooLarge':
-            if (page_params.max_file_upload_size > 0) {
-                // sanitization not needed as the file name is not potentially parsed as HTML, etc.
-                const context = {
-                    file_name: file.name,
-                    file_size: page_params.max_file_upload_size,
-                };
-                msg = i18n.t('"__file_name__" was too large; the maximum file size is __file_size__MB.',
-                             context);
-            } else {
-                // If uploading files has been disabled.
-                msg = i18n.t('File and image uploads have been disabled for this organization.');
-            }
-            break;
-        case 413: // HTTP status "Request Entity Too Large"
-            msg = i18n.t("Sorry, the file was too large.");
-            break;
-        case 400: {
-            const server_message = server_response && server_response.msg;
-            msg = server_message || i18n.t("An unknown error occurred.");
-            break;
+exports.setup_upload = function (config) {
+    const uppy = Uppy({
+        debug: true,
+        autoProceed: true,
+        restrictions: {
+            maxFileSize: page_params.max_file_upload_size * 1024 * 1024,
+        },
+        locale: {
+            strings: {
+                exceedsSize: i18n.t('This file exceeds maximum allowed size of'),
+                failedToUpload: i18n.t('Failed to upload %{file}'),
+            },
+        },
+    });
+    uppy.setMeta({
+        csrfmiddlewaretoken: csrf_token,
+    });
+    uppy.use(
+        XHRUpload, {
+            endpoint: '/json/user_uploads',
+            formData: true,
+            fieldName: 'file',
+            // Number of concurrent uploads
+            limit: 5,
+            locale: {
+                strings: {
+                    timedOut: i18n.t('Upload stalled for %{seconds} seconds, aborting.'),
+                },
+            },
         }
-        default:
-            msg = i18n.t("An unknown error occurred.");
-            break;
+    );
+
+    uppy.use(ProgressBar, {
+        target: exports.get_item("send_status_identifier", config),
+        hideAfterFinish: false,
```

Comment:
What's the thinking behind doing the hiding ourselves vs. using their built-in `hideAfterFinish` feature?

Reply:
hmm. Actually the issue was that when the file uploaded is small the progressbar used to disappear very quickly. So if a person has very high speed internet its not obvious what just happened. So we hide it manually 500ms after the upload is completed.

Reply:
OK, sounds good.  We can tune that 500ms after testing on chat.zulip.org; we usually use ~300ms for that purpose elsewhere.